qt/scientific_interfaces/ISISReflectometry/GUI/Common/Decoder.cpp

```diff
@@ -0,0 +1,463 @@
+  }
+  gui->m_ui.filterBox->setText(map[QString("filterBox")].toString());
+}
+
+void Decoder::decodeRunsTableModel(ReductionJobs *jobs,
+                                   const QList<QVariant> &list) {
+  for (auto groupIndex = 0; groupIndex < list.size(); ++groupIndex) {
+    auto group = decodeGroup(list[groupIndex].toMap());
+    jobs->mutableGroups()[groupIndex] = group;
+  }
+}
+
+MantidQt::CustomInterfaces::ISISReflectometry::Group
+Decoder::decodeGroup(const QMap<QString, QVariant> &map) {
+  auto rows = decodeRows(map["rows"].toList());
+  MantidQt::CustomInterfaces::ISISReflectometry::Group group(
+      map[QString("name")].toString().toStdString(), rows);
+  if (m_projectSave) {
+    auto itemState = map[QString("itemState")].toInt();
+    group.setState(State(itemState));
+  }
+  group.m_postprocessedWorkspaceName =
+      map[QString("postprocessedWorkspaceName")].toString().toStdString();
+  return group;
+}
+
+std::vector<boost::optional<MantidQt::CustomInterfaces::ISISReflectometry::Row>>
+Decoder::decodeRows(const QList<QVariant> &list) {
+  std::vector<
+      boost::optional<MantidQt::CustomInterfaces::ISISReflectometry::Row>>
+      rows;
+  for (const auto &rowMap : list) {
+    rows.emplace_back(decodeRow(rowMap.toMap()));
+  }
+  return rows;
+}
+
+namespace {
+ReductionOptionsMap decodeReductionOptions(const QMap<QString, QVariant> &map) {
+  ReductionOptionsMap rom;
+  for (const auto &key : map.keys()) {
+    rom.insert(std::pair<std::string, std::string>(
+        key.toStdString(), map[key].toString().toStdString()));
+  }
+  return rom;
+}
+} // namespace
+
+boost::optional<MantidQt::CustomInterfaces::ISISReflectometry::Row>
+Decoder::decodeRow(const QMap<QString, QVariant> &map) {
+  if (map.size() == 0) {
+    return boost::optional<
+        MantidQt::CustomInterfaces::ISISReflectometry::Row>();
+  }
+  std::vector<std::string> number;
+  for (const auto &runNumber : map[QString("runNumbers")].toList()) {
+    number.emplace_back(runNumber.toString().toStdString());
+  }
+  bool scaleFactorPresent = map[QString("scaleFactorPresent")].toBool();
+  double scaleFactor = 0;
+  if (scaleFactorPresent) {
+    scaleFactor = map[QString("scaleFactor")].toDouble();
+  }
+  if (scaleFactorPresent) {
+    auto row = MantidQt::CustomInterfaces::ISISReflectometry::Row(
+        number, map[QString("theta")].toDouble(),
+        decodeTransmissionRunPair(map[QString("transRunNums")].toMap()),
+        decodeRangeInQ(map[QString("qRange")].toMap()), scaleFactor,
+        decodeReductionOptions(map[QString("reductionOptions")].toMap()),
+        decodeReductionWorkspace(map[QString("reductionWorkspaces")].toMap()));
+    if (m_projectSave) {
+      auto itemState = map[QString("itemState")].toInt();
+      row.setState(State(itemState));
+      row.setOutputQRange(decodeRangeInQ(map[QString("qRangeOutput")].toMap()));
+    }
+    return row;
+  } else {
+    auto row = MantidQt::CustomInterfaces::ISISReflectometry::Row(
```

Comment:
There's quite a bit of duplication here - is the only difference whether we pass `scaleFactor` or `boost::none` when creating the row? If so, could that be turned into an optional prior to this so we only need to do all this once?

Reply:
That was the original implementation and would have been ideal, however, I have found that the compiler through errors when attempting to pass a `boost::optional<double>`

Reply:
This compiles ok for me...
```
  boost::optional<double> maybeScaleFactor = boost::none;
  bool scaleFactorPresent = map[QString("scaleFactorPresent")].toBool();
  if (scaleFactorPresent) {
    maybeScaleFactor = map[QString("scaleFactor")].toDouble();
  }
  auto row = MantidQt::CustomInterfaces::ISISReflectometry::Row(
      number, map[QString("theta")].toDouble(),
      decodeTransmissionRunPair(map[QString("transRunNums")].toMap()),
      decodeRangeInQ(map[QString("qRange")].toMap()), maybeScaleFactor,
      decodeReductionOptions(map[QString("reductionOptions")].toMap()),
      decodeReductionWorkspace(map[QString("reductionWorkspaces")].toMap()));
  if (m_projectSave) {
    auto itemState = map[QString("itemState")].toInt();
    row.setState(State(itemState));
    row.setOutputQRange(decodeRangeInQ(map[QString("qRangeOutput")].toMap()));
  }
```

Reply:
I don't remember what the exact issue was but that should work!

Reply:
The issue was GCC7 on RHEL7, Ubuntu Python 3 and Doctests builds servers have a warning about the `boost::optional<double> = boost::none;` declaration.

Reply:
We discussed this offline, but for the record, I think this is a spurious warning which you can get around by using `boost::make_optional` as per the second answer [here](https://stackoverflow.com/questions/21755206/how-to-get-around-gcc-void-b-4-may-be-used-uninitialized-in-this-funct)